File: backend/routes/sla.py
"""
Rutas para gestion de SLA (Service Level Agreement).

Sprint 4.5 - Endpoints de metricas y configuracion SLA.
"""

from flask import Blueprint, g, jsonify, request

from backend.core.roles import require_auth, require_role
from backend.services.sla_service import (
    actualizar_configuracion_sla,
    crear_configuracion_sla,
    eliminar_configuracion_sla,
    listar_configuraciones_sla,
    obtener_alertas_activas,
    obtener_metricas_sla,
    resolver_alerta_sla,
)


bp = Blueprint("sla", __name__, url_prefix="/api/sla")


# =============================================================================
# Metricas SLA (Dashboard)
# =============================================================================


@bp.route("/metricas", methods=["GET"])
@require_auth
def get_metricas_sla():
    """
    Obtener metricas de cumplimiento SLA.

    Query params:
        - periodo_dias: Dias hacia atras (default: 30)
        - por_criticidad: Incluir desglose por criticidad (default: false)

    Returns:
        - total_solicitudes: Total en el periodo
        - on_time: Cantidad a tiempo
        - warning: Cantidad con warning
        - breach: Cantidad incumplidas
        - porcentaje_cumplimiento: % de cumplimiento
        - por_criticidad: Desglose (si se solicita)
    """
    periodo_dias = request.args.get("periodo_dias", 30, type=int)
    por_criticidad = request.args.get("por_criticidad", "false").lower() == "true"

    metricas = obtener_metricas_sla(periodo_dias=periodo_dias, por_criticidad=por_criticidad)

    return jsonify({"ok": True, "data": metricas})


@bp.route("/alertas", methods=["GET"])
@require_auth
def get_alertas_activas():
    """
    Obtener alertas SLA activas.

    Query params:
        - solicitud_id: Filtrar por solicitud (opcional)
        - tipo: Filtrar por tipo (warning, breach, escalated)

    Returns:
        Lista de alertas activas
    """
    solicitud_id = request.args.get("solicitud_id", type=int)
    tipo = request.args.get("tipo")

    alertas = obtener_alertas_activas(solicitud_id=solicitud_id, tipo=tipo)

    return jsonify({"ok": True, "data": alertas, "total": len(alertas)})


@bp.route("/alertas/<int:alerta_id>/resolver", methods=["PUT", "POST"])
@require_auth
def resolver_alerta(alerta_id):
    """
    Resolver una alerta SLA.

    Path params:
        - alerta_id: ID de la alerta

    Returns:
        Resultado de la operacion
    """
    user_id = str(g.user.get("user_id", "system"))

    resultado = resolver_alerta_sla(alerta_id=alerta_id, resuelto_por=user_id)

    if resultado.get("resuelto"):
        return jsonify({"ok": True, "message": "Alerta resuelta correctamente"})
    else:
        return (
            jsonify(
                {
                    "ok": False,
                    "error": {"code": "not_found", "message": "Alerta no encontrada o ya resuelta"},
                }
            ),
            404,
        )


# =============================================================================
# CRUD Configuracion SLA (Admin)
# =============================================================================


@bp.route("/configuraciones", methods=["GET"])
@require_auth
def get_configuraciones():
    """
    Listar configuraciones SLA.

    Query params:
        - activo: Filtrar por estado activo (true/false)

    Returns:
        Lista de configuraciones SLA
    """
    activo_param = request.args.get("activo")
    activo = None
    if activo_param:
        activo = activo_param.lower() == "true"

    configs = listar_configuraciones_sla(activo=activo)

    return jsonify({"ok": True, "data": configs, "total": len(configs)})


@bp.route("/configuraciones", methods=["POST"])
@require_role(["admin", "gerente"])
def create_configuracion():
    """
    Crear nueva configuracion SLA.

    Body:
        - nombre: Nombre descriptivo
        - estado_desde: Estado inicial
        - estado_hasta: Estado destino
        - tiempo_objetivo_horas: Tiempo objetivo
        - criticidad: Criticidad (opcional)
        - descripcion: Descripcion (opcional)
        - tiempo_alerta_horas: Horas para alertar (opcional)
        - notificar_al_vencer: bool (default: true)
        - escalar_al_vencer: bool (default: false)
        - escalar_a_rol: Rol de escalamiento (opcional)

    Returns:
        ID de la configuracion creada
    """
    data = request.get_json(silent=True) or {}

    # Validar campos requeridos
    required = ["nombre", "estado_desde", "estado_hasta", "tiempo_objetivo_horas"]
    missing = [f for f in required if not data.get(f)]
    if missing:
        return (
            jsonify(
                {
                    "ok": False,
                    "error": {
                        "code": "validation_error",
                        "message": f"Campos requeridos faltantes: {', '.join(missing)}",
                    },
                }
            ),
            400,
        )

    user_id = str(g.user.get("user_id", "system"))

    resultado = crear_configuracion_sla(
        nombre=data["nombre"],
        estado_desde=data["estado_desde"],
        estado_hasta=data["estado_hasta"],
        tiempo_objetivo_horas=int(data["tiempo_objetivo_horas"]),
        criticidad=data.get("criticidad"),
        descripcion=data.get("descripcion"),
        tiempo_alerta_horas=(
            int(data["tiempo_alerta_horas"]) if data.get("tiempo_alerta_horas") else None
        ),
        notificar_al_vencer=data.get("notificar_al_vencer", True),
        escalar_al_vencer=data.get("escalar_al_vencer", False),
        escalar_a_rol=data.get("escalar_a_rol"),
        created_by=user_id,
    )

    return (
        jsonify(
            {"ok": True, "data": resultado, "message": "Configuracion SLA creada correctamente"}
        ),
        201,
    )


@bp.route("/configuraciones/<int:config_id>", methods=["PUT"])
@require_role(["admin", "gerente"])
def update_configuracion(config_id):
    """
    Actualizar configuracion SLA existente.

    Path params:
        - config_id: ID de la configuracion

    Body:
        Campos a actualizar (todos opcionales)

    Returns:
        Resultado de la operacion
    """
    data = request.get_json(silent=True) or {}

    if not data:
        return (
            jsonify(
                {
                    "ok": False,
                    "error": {
                        "code": "validation_error",
                        "message": "No hay campos para actualizar",
                    },
                }
            ),
            400,
        )

    resultado = actualizar_configuracion_sla(config_id, **data)

    if resultado.get("actualizado"):
        return jsonify({"ok": True, "message": "Configuracion SLA actualizada correctamente"})
    else:
        return (
            jsonify(
                {
                    "ok": False,
                    "error": {
                        "code": "not_found",
                        "message": resultado.get("mensaje", "Configuracion no encontrada"),
                    },
                }
            ),
            404,
        )


@bp.route("/configuraciones/<int:config_id>", methods=["DELETE"])
@require_role(["admin", "gerente"])
def delete_configuracion(config_id):
    """
    Eliminar (desactivar) configuracion SLA.

    Path params:
        - config_id: ID de la configuracion

    Returns:
        Resultado de la operacion
    """
    resultado = eliminar_configuracion_sla(config_id)

    if resultado.get("eliminado"):
        return jsonify({"ok": True, "message": "Configuracion SLA eliminada correctamente"})
    else:
        return (
            jsonify(
                {
                    "ok": False,
                    "error": {"code": "not_found", "message": "Configuracion no encontrada"},
                }
            ),
            404,
        )

File: SPMV3.0/frontend/src/services/sla.js
/**
 * Servicio API para SLA (Service Level Agreement)
 *
 * Endpoints:
 * - GET /api/sla/metricas - Metricas de cumplimiento
 * - GET /api/sla/alertas - Alertas activas
 * - PUT /api/sla/alertas/:id/resolver - Resolver alerta
 * - GET /api/sla/configuraciones - Listar configuraciones
 * - POST /api/sla/configuraciones - Crear configuracion (admin)
 * - PUT /api/sla/configuraciones/:id - Actualizar (admin)
 * - DELETE /api/sla/configuraciones/:id - Eliminar (admin)
 */

import api from './api'

/**
 * Obtener metricas de cumplimiento SLA
 * @param {Object} params
 * @param {number} params.periodoDias - Dias hacia atras (default: 30)
 * @param {boolean} params.porCriticidad - Incluir desglose por criticidad
 * @returns {Promise<Object>} Metricas SLA
 */
export const getMetricas = async ({ periodoDias = 30, porCriticidad = false } = {}) => {
  const response = await api.get('/sla/metricas', {
    params: {
      periodo_dias: periodoDias,
      por_criticidad: porCriticidad
    }
  })
  return response.data?.data || {}
}

/**
 * Obtener alertas SLA activas
 * @param {Object} params
 * @param {number} params.solicitudId - Filtrar por solicitud
 * @param {string} params.tipo - Filtrar por tipo (warning, breach, escalated)
 * @returns {Promise<Array>} Lista de alertas
 */
export const getAlertas = async ({ solicitudId, tipo } = {}) => {
  const response = await api.get('/sla/alertas', {
    params: {
      solicitud_id: solicitudId,
      tipo
    }
  })
  return response.data?.data || []
}

/**
 * Resolver una alerta SLA
 * @param {number} alertaId - ID de la alerta
 * @returns {Promise<boolean>} true si se resolvio
 */
export const resolverAlerta = async (alertaId) => {
  const response = await api.put(`/sla/alertas/${alertaId}/resolver`)
  return response.data?.ok === true
}

/**
 * Listar configuraciones SLA
 * @param {Object} params
 * @param {boolean} params.activo - Filtrar por estado activo
 * @returns {Promise<Array>} Lista de configuraciones
 */
export const getConfiguraciones = async ({ activo } = {}) => {
  const response = await api.get('/sla/configuraciones', {
    params: activo !== undefined ? { activo } : {}
  })
  return response.data?.data || []
}

/**
 * Crear nueva configuracion SLA (solo admin)
 * @param {Object} data
 * @param {string} data.nombre - Nombre descriptivo
 * @param {string} data.estadoDesde - Estado inicial
 * @param {string} data.estadoHasta - Estado destino
 * @param {number} data.tiempoObjetivoHoras - Tiempo objetivo
 * @param {string} data.criticidad - Criticidad (opcional)
 * @param {string} data.descripcion - Descripcion (opcional)
 * @param {number} data.tiempoAlertaHoras - Horas para alertar (opcional)
 * @param {boolean} data.notificarAlVencer - Notificar al vencer
 * @param {boolean} data.escalarAlVencer - Escalar al vencer
 * @param {string} data.escalarARol - Rol de escalamiento
 * @returns {Promise<Object>} Configuracion creada
 */
export const crearConfiguracion = async (data) => {
  const response = await api.post('/sla/configuraciones', {
    nombre: data.nombre,
    estado_desde: data.estadoDesde,
    estado_hasta: data.estadoHasta,
    tiempo_objetivo_horas: data.tiempoObjetivoHoras,
    criticidad: data.criticidad,
    descripcion: data.descripcion,
    tiempo_alerta_horas: data.tiempoAlertaHoras,
    notificar_al_vencer: data.notificarAlVencer ?? true,
    escalar_al_vencer: data.escalarAlVencer ?? false,
    escalar_a_rol: data.escalarARol
  })
  return response.data?.data
}

/**
 * Actualizar configuracion SLA (solo admin)
 * @param {number} configId - ID de la configuracion
 * @param {Object} data - Campos a actualizar
 * @returns {Promise<boolean>} true si se actualizo
 */
export const actualizarConfiguracion = async (configId, data) => {
  const payload = {}
  if (data.nombre !== undefined) payload.nombre = data.nombre
  if (data.estadoDesde !== undefined) payload.estado_desde = data.estadoDesde
  if (data.estadoHasta !== undefined) payload.estado_hasta = data.estadoHasta
  if (data.tiempoObjetivoHoras !== undefined) payload.tiempo_objetivo_horas = data.tiempoObjetivoHoras
  if (data.criticidad !== undefined) payload.criticidad = data.criticidad
  if (data.descripcion !== undefined) payload.descripcion = data.descripcion
  if (data.tiempoAlertaHoras !== undefined) payload.tiempo_alerta_horas = data.tiempoAlertaHoras
  if (data.notificarAlVencer !== undefined) payload.notificar_al_vencer = data.notificarAlVencer
  if (data.escalarAlVencer !== undefined) payload.escalar_al_vencer = data.escalarAlVencer
  if (data.escalarARol !== undefined) payload.escalar_a_rol = data.escalarARol
  if (data.activo !== undefined) payload.activo = data.activo

  const response = await api.put(`/sla/configuraciones/${configId}`, payload)
  return response.data?.ok === true
}

/**
 * Eliminar configuracion SLA (solo admin)
 * @param {number} configId - ID de la configuracion
 * @returns {Promise<boolean>} true si se elimino
 */
export const eliminarConfiguracion = async (configId) => {
  const response = await api.delete(`/sla/configuraciones/${configId}`)
  return response.data?.ok === true
}

// Export default con todos los metodos
export default {
  getMetricas,
  getAlertas,
  resolverAlerta,
  getConfiguraciones,
  crearConfiguracion,
  actualizarConfiguracion,
  eliminarConfiguracion
}
